# CatBoost Regressor

In [1]:
from pathlib import Path
import sys

import pandas as pd

from catboost import CatBoostRegressor

In [2]:
PROJECT_ROOT = Path.cwd().parent

if str(PROJECT_ROOT) not in sys.path:
    sys.path.append(str(PROJECT_ROOT))

In [3]:
from src.data_loader import load_data
from src.preprocessing import clean_data
from src.feature_engineering import engineer_features
from src.encoding import encode_features
from src.split_data import split_data
from src.evaluation import evaluate_model

## Load Dataset

In [4]:
df = load_data()

df = clean_data(df)

df = engineer_features(df)

df = encode_features(df)

X_train, X_test, y_train, y_test = split_data(df)

## Train CatBoost

In [5]:
cat = CatBoostRegressor(
    random_state=42,
    verbose=0,
)

In [6]:
cat.fit(
    X_train,
    y_train,
)

CatBoostRegressor(loss_function='RMSE', random_state=42, verbose=0)

In [7]:
y_pred = cat.predict(
    X_test
)

## Prediction Results

In [8]:
comparison = pd.DataFrame(
    {
        "Actual Price": y_test,
        "Predicted Price": y_pred,
    }
)

comparison.head(10)

,Actual Price,Predicted Price
1179,650.00,606.053900
342,716.00,804.421484
649,1584.00,1548.926197
772,1020.00,786.134146
803,1749.00,1576.987434
358,557.37,566.616569
44,999.00,937.448067
231,330.00,360.435816
618,2267.86,1894.580770
43,682.00,660.463609


## Evaluate Model

In [9]:
metrics = evaluate_model(
    y_test,
    y_pred,
)

metrics

{'MAE': 174.50576392445217,
 'RMSE': 254.08437425217508,
 'R2': 0.8699299601689575}

In [10]:
for key, value in metrics.items():
    print(f"{key}: {value:.4f}")

MAE: 174.5058
RMSE: 254.0844
R2: 0.8699


In [11]:
train_score = cat.score(
    X_train,
    y_train,
)

test_score = cat.score(
    X_test,
    y_test,
)

print(f"Train : {train_score:.4f}")

print(f"Test  : {test_score:.4f}")

Train : 0.9671
Test  : 0.8699


# Conclusion

CatBoost achieved excellent performance on the laptop price dataset.

Performance

- MAE ≈ 174.51
- RMSE ≈ 254.08
- R² ≈ 0.870

CatBoost became one of the strongest models in the project and performed almost identically to Gradient Boosting.

The final model selection will be performed after comparing all models.